### PR TITLE
feat: promote charts

### PR DESCRIPTION
**Pull request by rephus**
Closes: https://github.com/lightdash/lightdash/issues/9855

### Description:

[Screencast from 30-04-24 10:10:23.webm](https://github.com/lightdash/lightdash/assets/1983672/456fbc9d-8779-49c9-98de-26f868b701ca)

Creates chart and space 

[Screencast from 30-04-24 10:22:41.webm](https://github.com/lightdash/lightdash/assets/1983672/9982c0c8-6d09-4650-87dd-fab8a589d61d)

### Reviewer actions

- [ ] I have manually tested the changes in the preview environment
- [ ] I have reviewed the code
- [ ] I understand that "request changes" will block this PR from merging

### CODE REVIEW

**Comment by rephus** [comment]
This PR depends on https://github.com/lightdash/lightdash/pull/9875, merge that one first to make this easier to review

File: packages/frontend/src/components/DataOps/hooks/useUpstreamProject.ts [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+
+export const useUpdateMutation = (id: string) => {
+    const queryClient = useQueryClient();
+    const { showToastError, showToastSuccess } = useToaster();
+    return useMutation<null, ApiError, UpdateMetadata>(
+        (data) => updateProject(id, data),
+        {
+            mutationKey: ['project_update', id],
+            onSuccess: async () => {
+                await queryClient.invalidateQueries(['project', id]);
```
[the remark this answers is not in the document]
Comment (owlas) [reply packages/frontend/src/components/DataOps/hooks/useUpstreamProject.ts:21]: actually after testing this works well in the FE

File: packages/backend/src/services/ProjectService/ProjectService.ts [hunk 2]
```diff
@@ -4019,4 +4045,167 @@ export class ProjectService extends BaseService {
...
+                await checkPermissions(
+                    organizationUuid,
+                    upstreamProjectUuid,
+                    existingSpace[0],
+                    'the upstream space and project',
+                );
+                newSpaceUuid = existingSpace[0].uuid;
+            } else {
+                // We have 0 or more than 1 space with the same slug
+                await checkPermissions(
```
Comment (owlas) [inline packages/backend/src/services/ProjectService/ProjectService.ts:4145]: We should check if the user has access to create spaces in the upstream uproject

File: packages/backend/src/services/ProjectService/ProjectService.ts [hunk 3]
```diff
@@ -4019,4 +4045,167 @@ export class ProjectService extends BaseService {
...
+                upstreamChart.spaceUuid,
+            );
+
+            await checkPermissions(
+                organizationUuid,
+                upstreamProjectUuid,
+                upstreamSpace,
+                'the upstream chart and project',
+            );
+            const updatedChart = await this.savedChartModel.createVersion(
```
Comment (owlas) [inline packages/backend/src/services/ProjectService/ProjectService.ts:4194]: we should include name + description

File: packages/backend/src/services/ProjectService/ProjectService.ts [hunk 4]
```diff
@@ -3201,6 +3204,29 @@ export class ProjectService extends BaseService {
         );
     }
 
+    async updateMetadata(
```
[the remark this answers is not in the document]
Comment (rephus) [reply packages/backend/src/services/ProjectService/ProjectService.ts:3199]: done, I replaced this with `null` if we want to `unset` this . Undefined on knex by default ignores the field in `update`

File: packages/backend/src/services/DashboardService/DashboardService.test.ts [hunk 5]
```diff
@@ -131,19 +132,6 @@ describe('DashboardService', () => {
         );
 
         expect(result).toEqual({ ...dashboard, isPrivate: space.is_private });
-        expect(dashboardModel.create).toHaveBeenCalledTimes(1);
```
[the remark this answers is not in the document]
Comment (rephus) [reply packages/backend/src/services/DashboardService/DashboardService.test.ts]: reverted

**Review by owlas — CHANGES_REQUESTED** [review]
Code looks good but got this when promoting:

CleanShot 2024-05-07 at 10 59 44@2x

**Comment by rephus** [comment]
> Code looks good but got this when promoting:

Fixed error message 

Screenshot from 2024-05-07 15-38-13



tl;dr, we should never use `throw new Error` in the backend.

**Comment by owlas** [comment]
If you want the error to show correctly @rephus it should always be a child of `LightdashError` or it'll by default get wrapped in a `LightdashUnexpectedServerError`

**Review by owlas — APPROVED** [review]
